feat: pass in context for tracing

gruyaume opened the pull request:
## Overview

Ask for the `NrfDiscoveryQueryCb` callback method to contain a context instance.

## Context

This change necessary in the context of adding tracing support in the AMF (see https://github.com/omec-project/amf/pull/473). It allows to take the context that is built from individual NGAP calls and transfer it down to each API call the AMF makes to other network functions. Please review this change as a whole with the AMF change.

gruyaume commented on nrfcache/nrfcache.go:
```diff
@@ -347,13 +348,13 @@ func disableNrfCaching() {
 }
 
-func SearchNFInstances(nrfUri string, targetNfType, requestNfType models.NfType, param *Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (models.SearchResult, error) {
+func SearchNFInstances(nrfUri string, targetNfType, requestNfType models.NfType, param *Nnrf_NFDiscovery.SearchNFInstancesParamOpts, ctx context.Context) (models.SearchResult, error) {
```
Agreed, done!

gruyaume commented:
> I think this change will break the `nrfCache` in all NFs, except the `amf` when PR [omec-project/amf#473](https://github.com/omec-project/amf/pull/473) is merged. Are you planning to make changes to the other NFs, as well?

I implemented @ghislainbourgeois's suggestion which was to add a new `SearchNFInstancesWithContext`. This makes it so that we don't break the other NF's.

gab-arrobo reviewed (changes requested):
This PR should be transparent for the other NFs, so, I was trying to use it with the `ausf` but it fails with the following error:
```
 => ERROR [builder 5/5] RUN make all                                                                              19.4s
------
 > [builder 5/5] RUN make all:
1.589 Start building ausf....
1.589 cd .// && \
1.589 CGO_ENABLED=0 go build -o /go/src/ausf/bin/ausf ausf.go
18.18 # github.com/omec-project/ausf/service
18.18 service/init.go:140:70: cannot use consumer.SendNfDiscoveryToNrf (variable of type func(nrfUri string, targetNfType models.NfType, requestNfType models.NfType, param *Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (models.SearchResult, error)) as nrfcache.NrfDiscoveryQueryCb value in argument to nrfCache.InitNrfCaching
18.33 make: *** [Makefile:55: bin/ausf] Error 1
------
Dockerfile:24
--------------------
  22 |     WORKDIR $GOPATH/src/ausf
  23 |     COPY . .
  24 | >>> RUN make all
  25 |
  26 |     FROM alpine:3.22 AS ausf
--------------------
ERROR: failed to build: failed to solve: process "/bin/sh -c make all" did not complete successfully: exit code: 2
make: *** [Makefile:66: docker-build] Error 1
```

ghislainbourgeois reviewed (approved):
+1

gruyaume commented:
> > > > > > I think this change will break the `nrfCache` in all NFs, except the `amf` when PR [omec-project/amf#473](https://github.com/omec-project/amf/pull/473) is merged. Are you planning to make changes to the other NFs, as well?
> > > > > 
> > > > > 
> > > > > I implemented @ghislainbourgeois's suggestion which was to add a new `SearchNFInstancesWithContext`. This makes it so that we don't break the other NF's.
> > > > > Edit: It would still be a breaking change because the `NrfDiscoveryQueryCb` callback method expects a context. We could have another init method (ex. `InitNrfCachingWithContext`) which would take another callback method with context but it feels like we're growing the lib in an ugly way. @ghislainbourgeois do you have any suggestion here on how to avoid breaking the other NF's?
> > > > 
> > > > 
> > > > I would have a slight preference to go through the pain of updating all the NFs right away, mostly to make sure we do not have an in-between state for too long. The rest is mostly cosmetic and I would not worry too much about it.
> > > > WDYT? @gab-arrobo any preferences?
> > > 
> > > 
> > > I would prefer to go through the pain of updating the NFs right away
> > 
> > 
> > Fair enough, then I removed the SearchNFInstancesWithContext since we'll have to go through the refactoring in all other nf's anyway.
> 
> What is your timeline to make the changes in the NFs that get affected by this change? I am asking this because the Dependabot runs to update dependencies in the NFs on Sundays at 9 PM PST/PDT (typically runs between 9 and 10 PM) and if we merge this PR before we have a timeline for the NFs' updates, we will spend a lot of GitHub runners' time on GHAs that are going to fail due to the incompatibility of the nrfcache. So, I would like to merge this PR when we have a timeline for the NFs PRs and on a Sunday night (after 10 PM) or a Monday (This gives us about a week to review, test and merge the NFs PRs and avoid GHA failures).

I'm fine with making that work next week. I.E. All NF changes should be completed by July 18th.

gruyaume commented:
> > > What is your timeline to make the changes in the NFs that get affected by this change? I am asking this because the Dependabot runs to update dependencies in the NFs on Sundays at 9 PM PST/PDT (typically runs between 9 and 10 PM) and if we merge this PR before we have a timeline for the NFs' updates, we will spend a lot of GitHub runners' time on GHAs that are going to fail due to the incompatibility of the nrfcache. So, I would like to merge this PR when we have a timeline for the NFs PRs and on a Sunday night (after 10 PM) or a Monday (This gives us about a week to review, test and merge the NFs PRs and avoid GHA failures).
> > 
> > 
> > I'm fine with making that work next week. I.E. All NF changes should be completed by July 18th.
> 
> Sure, then, I am going to merge this PR this coming Sunday night or Monday morning.

I made the necessary PR's in all the NF's that were affected by this breaking change, which turned out to only be the ausf, pcf, smf, udm, and of course the amf, in the same PR where I am making the tracing changes. All the PR's are ready for review.

gab-arrobo commented:
> > Sure, then, I am going to merge this PR this coming Sunday night or Monday morning.
> 
> I made the necessary PR's in all the NF's that were affected by this breaking change, which turned out to only be the ausf, pcf, smf, udm, and of course the amf, in the same PR where I am making the tracing changes. All the PR's are ready for review.

Ok, thanks. I am going to review them this morning
